File: Spreadsheet/DependencyGraph/DependencyGraph.cs
// Skeleton implementation written by Joe Zachary for CS 3500, September 2013.
// Version 1.1 (Fixed error in comment for RemoveDependency.)
// Version 1.2 - Daniel Kopta 
//               (Clarified meaning of dependent and dependee.)
//               (Clarified names in solution/project structure.)

using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;

namespace SpreadsheetUtilities
{
    /**
       *  Diego Andino
       *  September 25, 2020
       */

    /// <summary>
    /// (s1,t1) is an ordered pair of strings
    /// t1 depends on s1; s1 must be evaluated before t1
    /// 
    /// A DependencyGraph can be modeled as a set of ordered pairs of strings.  Two ordered pairs
    /// (s1,t1) and (s2,t2) are considered equal if and only if s1 equals s2 and t1 equals t2.
    /// Recall that sets never contain duplicates.  If an attempt is made to add an element to a 
    /// set, and the element is already in the set, the set remains unchanged.
    /// 
    /// Given a DependencyGraph DG:
    /// 
    ///    (1) If s is a string, the set of all strings t such that (s,t) is in DG is called dependents(s).
    ///        (The set of things that depend on s)    
    ///        
    ///    (2) If s is a string, the set of all strings t such that (t,s) is in DG is called dependees(s).
    ///        (The set of things that s depends on) 
    //
    // For example, suppose DG = {("a", "b"), ("a", "c"), ("b", "d"), ("d", "d")}
    //     dependents("a") = {"b", "c"}
    //     dependents("b") = {"d"}
    //     dependents("c") = {}
    //     dependents("d") = {"d"}
    //     dependees("a") = {}
    //     dependees("b") = {"a"}
    //     dependees("c") = {"a"}
    //     dependees("d") = {"b", "d"}
    /// </summary>
    public class DependencyGraph
    {
        private Dictionary<string, HashSet<string>> dependents;
        private Dictionary<string, HashSet<string>> dependees;

        /// <summary>
        /// Creates an empty DependencyGraph.
        /// </summary>
        public DependencyGraph()
        {
            Size = 0;
            dependents = new Dictionary<string, HashSet<string>>();
            dependees = new Dictionary<string, HashSet<string>>();
        }


        /// <summary>
        /// The number of ordered pairs in the DependencyGraph.
        /// </summary>
        public int Size { get; private set; }


        /// <summary>
        /// The size of dependees(s).
        /// This property is an example of an indexer.  If dg is a DependencyGraph, you would
        /// invoke it like this:
        /// dg["a"]
        /// It should return the size of dependees("a")
        /// </summary>
        public int this[string s]
        {
            get
            {
                HashSet<string> res;

                if (dependees.TryGetValue(s, out res))
                    return res.Count;

                return 0;
            }
        }


        /// <summary>
        /// Reports whether dependents(s) is non-empty.
        /// </summary>
        public bool HasDependents(string s)
        {
            if (s == null)
                throw new ArgumentNullException(nameof(s), "Parameter cannot be null");

            if (dependents.ContainsKey(s))
            {
                if (dependents[s].Count > 0)
                    return true;
            }

            return false;
        }


        /// <summary>
        /// Reports whether dependees(s) is non-empty.
        /// </summary>
        public bool HasDependees(string s)
        {
            if (s == null)
                throw new ArgumentNullException(nameof(s), "Parameter cannot be null");

            if (dependees.ContainsKey(s))
            {
                if (dependees[s].Count > 0)
                    return true;
            }

            return false;
        }


        /// <summary>
        /// Enumerates dependents(s).
        /// </summary>
        public IEnumerable<string> GetDependents(string s)
        {
            if (s == null)
                throw new ArgumentNullException(nameof(s), "Parameter cannot be null");

            if (dependents.ContainsKey(s))
            {
                if (dependents[s].Any(item => item == "") && dependents[s].Count == 1)
                    return new HashSet<string>();

                else
                    return new HashSet<string>(dependents[s]);
            }
            
            return new HashSet<string>();
        }

        /// <summary>
        /// Enumerates dependees(s).
        /// </summary>
        public IEnumerable<string> GetDependees(string s)
        {
            if (s == null)
                throw new ArgumentNullException(nameof(s), "Parameter cannot be null");

            if (dependees.ContainsKey(s))
            {
                if (dependees[s].Any(item => item == "") && dependees[s].Count == 1)
                    return new HashSet<string>();

                else
                    return new HashSet<string>(dependees[s]);
            }

            return new HashSet<string>();
        }


        /// <summary>
        /// <para>Adds the ordered pair (s,t), if it doesn't exist</para>
        /// 
        /// <para>This should be thought of as:</para>   
        /// 
        ///   t depends on s
        ///
        /// </summary>
        /// <param name="s"> s must be evaluated first. T depends on S</param>
        /// <param name="t"> t cannot be evaluated until s is</param>        /// 
        public void AddDependency(string s, string t)
        {
            if (s == null || t == null)
                throw new ArgumentNullException("Key or value cannot be null");

            if (dependents.ContainsKey(s))
            {
                if (dependents[s].Contains(t))
                    return;
                
                dependents[s].Add(t);
                AddDependee(s, t);
                
                Size++;
            }

            else
            {
                dependents.Add(s, new HashSet<string>(){t});
                AddDependee(s, t);

                Size++;
            }
        }


        /// <summary>
        /// Private helper method to add dependees.
        /// </summary>
        /// <param name="s"></param>
        /// <param name="t"></param>
        private void AddDependee(string s, string t)
        {
            if (!dependees.ContainsKey(t))
                dependees.Add(t, new HashSet<string>(){s});
            
            else
                dependees[t].Add(s);
        }


        /// <summary>
        /// Removes the ordered pair (s,t), if it exists
        /// </summary>
        /// <param name="s"></param>
        /// <param name="t"></param>
        public void RemoveDependency(string s, string t)
        {
            if (s == null || t == null)
                throw new ArgumentNullException("Key or value cannot be null");

            HashSet<string> deps;

            if (dependents.TryGetValue(s, out deps))
            {
                if (deps.Remove(t))
                    Size--;
            }

            if (dependees.TryGetValue(t, out deps))
                deps.Remove(s);

            if (dependents.ContainsKey(s) && dependents[s].Count == 0)
                dependents.Remove(s);

            if (dependents.ContainsKey(t) && dependents[t].Count == 0)
                dependees.Remove(t);
        }


        /// <summary>
        /// Removes all existing ordered pairs of the form (s,r).  Then, for each
        /// t in newDependents, adds the ordered pair (s,t).
        /// </summary>
        public void ReplaceDependents(string s, IEnumerable<string> newDependents)
        {
            if (s == null || newDependents == null)
                throw new ArgumentNullException("The key or IEnumerable cannot be null.");

            foreach (string r in GetDependents(s))
                RemoveDependency(s, r);
            
            foreach (string t in newDependents)
                AddDependency(s, t);
        }


        /// <summary>
        /// Removes all existing ordered pairs of the form (r,s).  Then, for each 
        /// t in newDependees, adds the ordered pair (t,s).
        /// </summary>
        public void ReplaceDependees(string s, IEnumerable<string> newDependees)
        {
            if (s == null || newDependees == null)
                throw new ArgumentNullException("The key or IEnumerable cannot be null.");

            foreach (string r in GetDependees(s))
                RemoveDependency(r, s);

            foreach (string t in newDependees)
                AddDependency(t, s);
        }
    }
}

File: Server/Server/dependency_graph.cpp
#include "dependency_graph.h"
#include <stdexcept> // for exeptions

//this is mainly a translation from my own code in 3500

DependencyGraph::DependencyGraph()
{
	size = 0;
}

int DependencyGraph::get_size()
{
	return size;
}

/// <summary>
/// checks if a key has dependents, returns false if the key does not exist
/// </summary>
/// <param name="s">string to search for</param>
/// <returns>if a key has dependents</returns>
bool DependencyGraph::has_dependents(std::string& s)
{
	try
	{
		std::unordered_set<std::string> set = dependents.at(s);
		return set.size() > 0;
	}
	catch (std::out_of_range& e)
	{
		return false;
	}

}


/// <summary>
/// checks if a key has dependees, returns false if the key does not exist
/// </summary>
/// <param name="s">string to search for</param>
/// <returns>if a key has dependees</returns>
bool DependencyGraph::has_dependees(std::string& s)
{
	try
	{
		std::unordered_set<std::string> set = dependees.at(s);
		return set.size() > 0;
	}
	catch (std::out_of_range& e)
	{
		return false;
	}
}


/// <summary>
/// Enumerates dependents(s).
/// </summary>
std::vector<std::string> DependencyGraph::get_dependents(std::string& s)
{
	std::vector<std::string> ret;

	if (has_dependents(s))
	{
		for (std::string str : dependents.at(s))
		{
			ret.push_back(str);
		}
	}

	return ret;
}


std::vector<std::string> DependencyGraph::get_dependees(std::string& s)
{
	std::vector<std::string> ret;

	if (has_dependees(s))
	{
		for (std::string str : dependees.at(s))
		{
			ret.push_back(str);
		}
	}

	return ret;
}

/// <summary>
/// <para>Adds the ordered pair (s,t), if it doesn't exist</para>
/// 
/// <para>This should be thought of as:</para>   
/// 
///   t depends on s
///
/// </summary>
/// <param name="s"> s must be evaluated first. T depends on S</param>
/// <param name="t"> t cannot be evaluated until s is</param>        /// 
void DependencyGraph::add_dependency(std::string& s, std::string& t)
{
	bool ee = false;
	bool ent = false;

	//first takes care of the dependent side of things
	if (dependents.contains(s))
	{
		int ent_size = dependents.size();
		dependents.at(s).insert(t);

		//ent = ent_size != dependents.size();
		ent = true;
	}
	else
	{
		//creates a new HashSet for the dependents of s and adds t to it
		std::unordered_set<std::string> s_dependents;
		s_dependents.insert(t);
		//creates a new key, s, and maps it to the HashSet containing t
		dependents.insert({ s, s_dependents }); //don't ask me about this syntax, C++ is wild
		ent = true;
	}

	//second takes care of the dependee side of things
	if (dependees.contains(t))
	{
		int ee_size = dependees.size();
		dependees.at(t).insert(s);

		//ee = dependees.at(t).size() != ee_size;
		ee = true;
	}
	else
	{
		std::unordered_set<std::string> t_dependees;
		t_dependees.insert(s);
		dependees.insert({ t, t_dependees });
		ee = true;
	}

	//increments the size only if something new was added
	if (ent && ee)
		size++;
}

/// <summary>
/// Removes the ordered pair (s,t), if it exists
/// </summary>
/// <param name="s"></param>
/// <param name="t"></param>
void DependencyGraph::remove_dependency(std::string& s, std::string& t)
{

	if (dependents.contains(s) && dependees.contains(t))
	{		
		dependents.erase(t);
		dependees.erase(s);

		if (dependents.contains(s) && dependents.at(s).size() == 0)
		{
			dependents.erase(s);
		}

		if (dependees.contains(t) && dependees.at(t).size() == 0)
		{
			dependees.erase(t);
		}
		
		size--;

	}

	//it was originally here
	//size--;
}

void DependencyGraph::replace_dependents(std::string& s, std::vector<std::string>& new_dependents)
{
	//removes all the dependcies related to s' dependents
	for (std::string str : get_dependents(s))
	{
		remove_dependency(s, str);
	}

	//adds all new nodes from newDependents as dependencies
	for(std::string str : new_dependents)
	{
		add_dependency(s, str);
	}
}

/// <summary>
/// The size of dependees(s).
/// This property is an example of an indexer.  If dg is a DependencyGraph, you would
/// invoke it like this:
/// dg["a"]
/// It should return the size of dependees("a")
/// </summary>
int DependencyGraph::operator[](std::string s)
{
	return get_dependees(s).size();
}
